### PR TITLE
Added flow for creating file ZIP archive

### PR DESCRIPTION
**Pull request by michalbogacz**
###  Purpose

Flow for creating ZIP archive from files.
Example usage: download many S3 files as one ZIP file

It does not fully address #658 because unzipping is not here.

## References
References #658

## Changes
- Added new API `Archive` in `File` connector (with implementation and tests)

## Background Context
- for creating ZIP archive Java ZipOutputStream is used
- ZipOutputStream logic is in one stage 
- since stage containing zipping logic requires information when input files starts and ends, special `ByteString`s are inserted

### CODE REVIEW

File: file/src/main/scala/akka/stream/alpakka/file/model.scala [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+ */
+
+package akka.stream.alpakka.file
+
+final class ArchiveMetadata private (
+    val filePath: String
+)
+
+object ArchiveMetadata {
+  def create(filePath: String): ArchiveMetadata = new ArchiveMetadata(filePath)
```
Comment (2m) [inline file/src/main/scala/akka/stream/alpakka/file/model.scala]: Add an `apply` method here as well.

File: file/src/test/scala/docs/scaladsl/ArchiveSpec.scala [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+         val fileStream1: Source[ByteString,  Any] = ...
+         val fileStream2: Source[ByteString,  Any] = ...
+
+         // #sample
+         */
+
+        // #sample
+        val filesStream = Source(
+          List(
+            (ArchiveMetadata.create("akka_full_color.svg"), fileStream1),
```
Comment (2m) [inline file/src/test/scala/docs/scaladsl/ArchiveSpec.scala]: Replace with `ArchiveMetadata(...)`

File: file/src/main/scala/akka/stream/alpakka/file/javadsl/Archive.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ */
+object Archive {
+
+  /**
+   * Flow for compressing multiple files into one ZIP file.
+   */
+  def zipFlow(): Flow[Pair[ArchiveMetadata, Source[ByteString, NotUsed]], ByteString, NotUsed] =
+    Flow
+      .create[Pair[ArchiveMetadata, Source[ByteString, NotUsed]]]()
+      .map(pair => (pair.first, pair.second.asScala))
```
Comment (2m) [inline file/src/main/scala/akka/stream/alpakka/file/javadsl/Archive.scala]: This line fails to compile in Scala 2.11. Please help the compiler by wrapping the function with `func` as it is done in other connectors: https://github.com/akka/alpakka/blob/261645f14b6a99e36c8f61c7887f878c7883b563/s3/src/main/scala/akka/stream/alpakka/s3/javadsl/S3.scala#L392

File: file/src/test/java/docs/javadsl/ArchiveTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    CompletionStage<IOResult> ioResult = source.via(Archive.zipFlow()).runWith(fileSink, mat);
+
+    // #sample
+
+    ioResult.toCompletableFuture().get(3, TimeUnit.SECONDS);
+
+    archiveHelper.createReferenceZipFile(Arrays.asList(filePath1, filePath2), "logo-reference.zip");
+
+    final Sink<ByteString, CompletionStage<ByteString>> foldSink =
+        Sink.fold(ByteString.empty(), ByteString::concat);
```
Comment (2m) [inline file/src/test/java/docs/javadsl/ArchiveTest.java]: `ByteString.empty()` should be replaced to `ByteString.emptyByteString()` so it compiles with Scala 2.13.

**Comment by 2m** [comment]
One of the validation tasks failed, because the master at the time you branched off, had a fatal warning. Please rebase this PR on top of the latest master to fix it.

**Comment by 2m** [comment]
This is a continuation of https://github.com/akka/akka-stream-contrib/pull/171

File: file/src/main/scala/akka/stream/alpakka/file/javadsl/Archive.scala (hunk 3 above)
Comment (michalbogacz) [reply file/src/main/scala/akka/stream/alpakka/file/javadsl/Archive.scala]: Thanks! I was struggling a little with it

**Comment by michalbogacz** [comment]
@2m Sorry, I messed with my fork. Is it ok, if I create new PR for this?

**Comment by ennru** [comment]
Sure, go ahead with a new PR and close this one.

**Comment by 2m** [comment]
You can always fix-up your branch and force push it. That will update this PR accordingly,

But, of course, you can always create a new branch and open a new PR from that.

**Comment by michalbogacz** [comment]
I force pushed code once again. @2m review fixes are included. 
If you see some other improvements, let me know :)

**Comment by mkurz** [comment]
Any chance to also implement unzip?

**Pull request merged**